src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs

```diff
@@ -339,19 +338,60 @@ public byte[] Http2AltSvcOriginUri
         private ValueTask<(HttpConnectionBase? connection, bool isNewConnection, HttpResponseMessage? failureResponse)>
             GetConnectionAsync(HttpRequestMessage request, bool async, CancellationToken cancellationToken)
         {
-            if (_http3Enabled && request.Version.Major >= 3)
+            // Do not even attempt at getting/creating a connection if it's already obvious we cannot provided the one requested.
+            if (request.VersionPolicy != HttpVersionPolicy.RequestVersionOrLower)
+            {
+                if (request.Version.Major >= 3 && !_http3Enabled)
+                {
+                    return ValueTask.FromException<(HttpConnectionBase? connection, bool isNewConnection, HttpResponseMessage? failureResponse)>
+                        (new HttpRequestException(SR.Format(SR.net_http_requested_version_not_enabled, request.Version, request.VersionPolicy, 3)));
+                }
+                if (request.Version.Major >= 2 && !_http2Enabled)
```

Comment:
What happens if requested version is HTTP/3, _http3Enabled is true, but _http2Enabled is false?  This is going to throw; that's desirable?

Reply:
I see what you're pointing at, there should be `else if` instead of just `if`, I'll fix it.
However, with how we implemented [`HttpConnectionSettings._maxHttpVersion`](https://github.com/dotnet/runtime/blob/master/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionSettings.cs#L70-L74), it's not possible to have H3 enabled while H2 disabled.

Reply:
Fixed.